### ISSUE TITLE
"image_pull_policy" environment variable not respected

### ISSUE THREAD

**Issue by dedene** [openfaas-incubator/openfaas-operator#56]
## Expected Behaviour
OpenFaaS is deployed using Faas-netes and Helm, with values

```
operator:
  image: openfaas/openfaas-operator:0.8.9
  create: true
  createCRD: true
```

checking that the image_pull_policy is set for the operator container in the gateway pod:
```
$ kubectl -n openfaas describe deployment gateway

[...]
   operator:
    Image:  openfaas/openfaas-operator:0.8.9
    Port:   8081/TCP
    Command:
      ./openfaas-operator
      -logtostderr
      -v=2
    Environment:
      function_namespace:  openfaas-fn
      read_timeout:        60s
      write_timeout:       60s
      image_pull_policy:   Always
[...]
```

and creating a function where the docker tag is not "latest", i.e. the echo example from https://github.com/openfaas/faas/blob/master/sample-functions/echo/Dockerfile with following definition:

```
provider:
  name: faas  
functions:
  echo:
    lang: dockerfile
    handler: ./echo
    image: dedene/echo:v1
```

I was expecting the deployment to have "imagePullPolicy: Always" 

## Current Behaviour

The deployment has the imagePullPolicy: IfNotPresent

```
$ kubectl get deployment echo -o yaml

[...]
   spec:
      containers:
      - image: dedene/echo:v1
        imagePullPolicy: IfNotPresent
[...]
```

When OpenFaaS is deployed using Faas-netes, with

```
operator:
  create: false
```

the resulting deployment does have the correct imagePullPolicy.

## Possible Solution

It seems when using the :latest tag, the imagePullPolicy does get set to "Always". So it follows the behavior specified in https://kubernetes.io/docs/concepts/configuration/overview/#container-images when "imagePullPolicy" is omitted. It could be the imagePullPolicy does not get applied correctly in the deployment by the operator?

## Steps to Reproduce (for bugs)
1. deploy OpenFaaS using openfaas/openfaas-operator:0.8.9
2. deploy any function where you specify a tag different than :latest on the docker image
3. inspect the resulting deployment or pod specification

## Your Environment
* Docker version `docker version`: Docker Version 18.09.0
* What version and distriubtion of Kubernetes are you using? `kubectl version`: Could reproduce on both v1.10.3 and v1.12.1
* Operating System and version (e.g. Linux, Windows, MacOS): MacOS 10.14.1, but not relevant for this bug
* What network driver are you using and what CIDR? Could reproduce on K8s 1.12 with Weave and 1.10 with Flannel

**Comment by alexellis** [comment]
@LucasRoesler PTAL

**Comment by LucasRoesler** [comment]
**TL;DR:** This value should be `Always` by default but is/can be configured via the `image_pull_policy` ENV variable in the operator pod.  When deploying with the Helm chart this is controlled by the `faasnetesd.imagePullPolicy` value.

@dedene can you verify the your helm value and the value of the ENV variable?

---

**The Gory Details**

I dug through the code a little bit so that I could document the flow for how that "image pull policy" is set for the function.  

There are a couple of points in the code worth pointing out.

First, the method that actually creates the function Deployment in k8s accepts the `imagePullPolicy` as an argument, whatever this value is will be the value you see on the function pods

https://github.com/openfaas-incubator/openfaas-operator/blob/10f0095dd725a1359acec0e89253bdfc8318d9b8/pkg/controller/deployment.go#L25-L28

This method is used during Create/Update of the Function object and it uses the `imagePullPolicy` value defined on the `Controller` object, that is the `c.imagePullPolicy` on this line:

https://github.com/openfaas-incubator/openfaas-operator/blob/10f0095dd725a1359acec0e89253bdfc8318d9b8/pkg/controller/controller.go#L332

So, now we need to find where this is set, 

https://github.com/openfaas-incubator/openfaas-operator/blob/10f0095dd725a1359acec0e89253bdfc8318d9b8/pkg/controller/controller.go#L86-L91

Which is called in the `main` 

https://github.com/openfaas-incubator/openfaas-operator/blob/10f0095dd725a1359acec0e89253bdfc8318d9b8/main.go#L88

Which ultimately is set here

https://github.com/openfaas-incubator/openfaas-operator/blob/10f0095dd725a1359acec0e89253bdfc8318d9b8/main.go#L72-L78

Reading the code this value should be `Always` by default but can be configured via the `image_pull_policy` ENV variable in the operator pod.  When deploying with the Helm chart this is controlled by the `faasnetesd.imagePullPolicy` value.

**Comment by LucasRoesler** [comment]
Right now, just waiting on the verification of the env value from Peter

**Comment by wouterh** [comment]
@LucasRoesler I am a colleague of @dedene:

the environment variable is set correctly:

```
$ kubectl get pods gateway-8fdf48889-4tcft -o yaml | grep -B 7 -A 2 image_pull_policy
    env:
    - name: function_namespace
      value: openfaas-fn
    - name: read_timeout
      value: 60s
    - name: write_timeout
      value: 60s
    - name: image_pull_policy
      value: Always
    image: openfaas/openfaas-operator:0.8.9
```

To verify even more:

```
$ kubectl exec gateway-8fdf48889-4tcft -c operator -- sh -c 'echo $image_pull_policy'
Always
```

For a freshly deployed function:

```
$ kubectl get pods -n openfaas-fn <<REDACTED>> -o jsonpath="{.spec.containers[0].imagePullPolicy}"
IfNotPresent
```

**Comment by LucasRoesler** [comment]
@wouterh thanks for following up.  Unrelated to this issue, i noticed that you are using operator version 0.8.9, the latest release is 0.9.1.  There is a small chance, but perhaps upgrading might help.

I have to be honest, I am at a loss.  I reviewed the code again, and it seems like it should (in the worst case) default to Always.  I will have to hack on it a little bit to trace it a little better.

Perhaps @stefanprodan could review this as well and will see something that I missed?

**Comment by wouterh** [comment]
I also reviewed the related code and I am also puzzled as to why it doesn't work as expected. I will try upgrading later today 🤞

**Comment by wouterh** [comment]
@LucasRoesler I've reviewed the code again and I think I spotted the error. The imagePullPolicy passed as an argument to `NewController` is never copied into the controller structure:

https://github.com/openfaas-incubator/openfaas-operator/blob/master/pkg/controller/controller.go#L110-L119

**Comment by LucasRoesler** [comment]
Nice catch and PR!
